### PR TITLE
core/vm: cherry-pick Shanghai hardfork from go-ethereum

### PR DESCRIPTION
**Pull request by minh-bq**
This PR cherry-picks 3 Shanghai EIPs
- EIP-3651 - Starts the COINBASE address warm. Commit: [ec2ec2d08e](https://github.com/ethereum/go-ethereum/commit/ec2ec2d08e) ("core: implement EIP-3651, warm coinbase")
- EIP-3855 - New PUSH0 instruction. Commit: [3b967d16ca](https://github.com/ethereum/go-ethereum/commit/3b967d16ca) ("core/vm: implement EIP-3855: PUSH0 instruction")
- EIP-3860 - Limit and meter initcode. Commit: [793f0f9ec8](https://github.com/ethereum/go-ethereum/commit/793f0f9ec8) ("core/vm: implement EIP-3860: Limit and meter initcode")

This also cherry-picks 
- EIP-1153 - Transient storage opcodes. Commit: [b4ea2bf7dd](https://github.com/ethereum/go-ethereum/commit/b4ea2bf7dd) ("all: implement EIP-1153 transient storage")
since this commit is a prerequisite of the above commits. However, this EIP is not enabled in Shanghai.

This PR also includes a commit to increase max code size and max initcode size to 32768 and 65536 respectively after Shanghai.

### CODE REVIEW

**Review by andicrypt — COMMENTED** [review]
LGTM

**Review by andicrypt — APPROVED** [review]

File: core/state/statedb.go [hunk 1]
```diff
@@ -892,9 +928,10 @@ func (s *StateDB) IntermediateRoot(deleteEmptyObjects bool) common.Hash {
 	return s.trie.Hash()
 }
 
-// Prepare sets the current transaction hash and index which are
-// used when the EVM emits new state logs.
-func (s *StateDB) Prepare(thash common.Hash, ti int) {
+// SetTxContext sets the current transaction hash and index which are
```
Comment (huyngopt1994) [inline core/state/statedb.go:931]: sets the current transaction hash, index and accessList

File: params/protocol_params.go [hunk 2]
```diff
@@ -127,8 +127,9 @@ const (
 	ElasticityMultiplier     = 2          // Bounds the maximum gas limit an EIP-1559 block may have.
 	InitialBaseFee           = 1000000000 // Initial base fee for EIP-1559 blocks.
 
-	MaxCodeSize     = 24576           // Maximum bytecode to permit for a contract
-	MaxInitCodeSize = 2 * MaxCodeSize // Maximum initcode to permit in a creation transaction and create instructions
+	MaxCodeSize         = 24576 // Maximum bytecode to permit for a contract
+	MaxCodeSizeShanghai = 32768
```
Comment (huyngopt1994) [inline params/protocol_params.go:131]: Should we apply comment here `// Maximum bytecode to permit for a contract after Shanghai Hardfork`

File: core/state/statedb.go (hunk 1 above)
Comment (minh-bq) [reply core/state/statedb.go:931]: This is nit and the comment is from go-ethereum so a fix can make future conflict

File: params/protocol_params.go (hunk 2 above)
Comment (minh-bq) [reply params/protocol_params.go]: added

File: core/vm/eips.go [hunk 3]
```diff
@@ -27,6 +27,7 @@ import (
 
 var activators = map[int]func(*JumpTable){
 	3855: enable3855,
+	3860: enable3860,
```
Comment (huyngopt1994) [inline core/vm/eips.go:30]: hm although it's unorder keys, why we leave 3860 behind  3855

File: core/tx_pool.go [hunk 4]
```diff
@@ -263,6 +263,7 @@ type TxPool struct {
 	odysseus bool // Fork indicator whether we are in the Odysseus stage.
 	antenna  bool // Fork indicator whether we are in Antenna stage.
 	miko     bool // Fork indicator whether we are using sponsored trnasactions.
```
Comment (huyngopt1994) [inline core/tx_pool.go:266]: trnasactions -> transactions

**Review by huyngopt1994 — APPROVED** [review]

**Pull request merged**